packages/pubsub/src/subscription.js

```diff
@@ -592,30 +365,249 @@ Subscription.prototype.decorateMessage_ = function(message) {
+      }
+    }
+  });
+
+  this.on('removeListener', function(event) {
+    if (event === 'message' && --self.messageListeners === 0) {
+      self.closeConnection_();
+    }
+  });
+};
+
+/**
+ *
+ */
+Subscription.prototype.createMessage_ = function(data) {
+  var message = new Message(self, data);
+  this.messages_.add(message);
+  return message;
+};
+
+/**
+ *
+ */
+Subscription.prototype.modifyAckDeadline_ = function(messages, callback) {
+  var self = this;
+  var reqOpts;
+
+  if (this.connection) {
+    reqOpts = messages.reduce(function(reqOpts, message) {
+      message.modifyDeadlineAckIds.push(message.ackId);
+      message.modifyDeadlineSeconds.push(message.deadline);
+      return reqOpts;
+    }, {
+      modifyDeadlineAckIds: [],
+      modifyDeadlineSeconds: []
+    });
+
+    this.connection.write(reqOpts);
+    setImmediate(onComplete);
+    return;
+  }
+
+  var requests = groupByDeadline(messages).map(function(reqOpts) {
+    return self.request({
+      client: 'subscriberClient',
+      method: 'modifyAckDeadline',
+      reqOpts: reqOpts
+    }).catch(function(err) {
+      self.emit('error', err);
+    });
+  });
+
+  Promise.all(requests)
+    .then(onComplete.bind(null, null), onComplete);
+
+  function groupByDeadline(messages) {
+    var requests = new Map();
+
+    messages.forEach(function(message) {
+      if (!requests.has(message.deadline)) {
+        requests.set(message.deadline, {
+          subscription: self.name,
+          ackIds: [],
+          ackDeadlineSeconds: message.deadline
+        });
+      }
+
+      var request = requests.get(message.deadline);
+      request.ackIds.push(message.ackId);
+    });
+
+    return Array.from(requests.values());
+  }
+
+  function onComplete(err, resp) {
+    messages.forEach(function(message) {
+      if (message.deadline === 0) {
+        self.messages_.delete(message);
+      }
+    });
+
+    if (self.connection.isPaused()) {
+      if (self.messages_.size < self.flowControl.maxMessages) {
+        self.connection.resume();
+      }
+    }
+
+    callback(err, resp);
+  }
+};
+
+/**
+ * @param {object} config - The push config.
+ * @param {string} config.pushEndpoint
+ * @param {object} config.attributes
+ */
+Subscription.prototype.modifyPushConfig = function(config, gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
+
+  var reqOpts = {
+    subscription: this.name,
+    pushConfig: config
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'modifyPushConfig',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
+
+/**
+ * @private
+ */
+Subscription.prototype.openConnection_ = function() {
+  var self = this;
+
+  var reqOpts = {
+    subscription: this.name,
+    streamAckDeadlineSeconds: this.ackDeadline
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'streamingPull',
+    returnFn: true,
+    reqOpts: {
+      highWaterMark: self.flowControl.highWaterMark
```

Comment:
> One thing to note is that I renamed maxBytes to highWaterMark as it is the standard configuration name in Node.js Streams.

This is not going to work. This is not a replacement for maxBytes.

- Buffers are bad. We don't want to buffer any messages inside the Stream object. A message in a buffer is just wasted space and time (it's memory consumed by a message that isn't being processed by the user). Just pull from the stream aggressively, until the user is flow controlled (at which point the stream needs to be paused, as you are doing), and deliver everything immediately. 
- Also, it misses the point of flow control. We are trying to protect the user from accumulating too many messages in memory (and crashing the process). They can still run out of memory this way since this is only protecting the buffer, not the messages delivered to the event handler. You need to keep track of the number of messages and total size of messages in the current inventory, and pause when the inventory goes over limit. That's what maxMessages and maxBytes are for. This might be done correctly for maxMessages already. It needs to be done for maxBytes as well.